File: apps/mobile/lib/features/diagnostic/domain/diagnostic.dart
enum DiagnosticConfidence {
  guessing('guessing', 'Guessing'),
  somewhatSure('somewhat_sure', 'Somewhat sure'),
  verySure('very_sure', 'Very sure');

  const DiagnosticConfidence(this.value, this.label);
  final String value;
  final String label;
}

class DiagnosticOption {
  DiagnosticOption.fromJson(Map<String, dynamic> json)
    : id = json['id'] as String,
      label = json['label'] as String;
  final String id;
  final String label;
}

class DiagnosticQuestion {
  DiagnosticQuestion.fromJson(Map<String, dynamic> json)
    : id = json['id'] as String,
      category = json['category'] as String,
      interactionType = json['interactionType'] as String,
      skill = (json['skill'] as Map<String, dynamic>)['label'] as String,
      concept = (json['concept'] as Map<String, dynamic>)['label'] as String,
      prompt = json['prompt'] as String,
      context = json['context'] as String?,
      code = json['code'] as String?,
      options = List.unmodifiable(
        (json['options'] as List).map(
          (item) => DiagnosticOption.fromJson(item as Map<String, dynamic>),
        ),
      ),
      confidenceRequested = json['confidenceRequested'] as bool {
    if (interactionType != 'single_choice' ||
        options.length < 2 ||
        options.map((option) => option.id).toSet().length != options.length) {
      throw const FormatException('Unsupported diagnostic question');
    }
  }
  final String id;
  final String category;
  final String interactionType;
  final String skill;
  final String concept;
  final String prompt;
  final String? context;
  final String? code;
  final List<DiagnosticOption> options;
  final bool confidenceRequested;
  String get categoryLabel => switch (category) {
    'scenario_judgment' => 'Scenario judgment',
    'predict_outcome' => 'Predict the outcome',
    'spot_the_bug' => 'Spot the bug',
    'better_approach' => 'Choose a better approach',
    'conceptual_reasoning' => 'Conceptual reasoning',
    _ => 'Engineering diagnostic',
  };
}

class DiagnosticReview {
  DiagnosticReview.fromJson(Map<String, dynamic> json)
    : question = DiagnosticQuestion.fromJson(
        json['question'] as Map<String, dynamic>,
      ),
      selectedOptionId = json['selectedOptionId'] as String,
      correct = json['correct'] as bool,
      correctOptionId = json['correctOptionId'] as String,
      explanation = json['explanation'] as String,
      keyIdea = json['keyIdea'] as String;
  final DiagnosticQuestion question;
  final String selectedOptionId;
  final bool correct;
  final String correctOptionId;
  final String explanation;
  final String keyIdea;
  String get correctAnswer => question.options
      .firstWhere((option) => option.id == correctOptionId)
      .label;
}

class DiagnosticSnapshot {
  DiagnosticSnapshot.fromJson(Map<String, dynamic> json)
    : completed = json['status'] == 'completed',
      answered = (json['progress'] as Map<String, dynamic>)['answered'] as int,
      total = (json['progress'] as Map<String, dynamic>)['total'] as int,
      question = json['question'] == null
          ? null
          : DiagnosticQuestion.fromJson(
              json['question'] as Map<String, dynamic>,
            ),
      confidenceQuestion = json['confidence'] == null
          ? null
          : DiagnosticQuestion.fromJson(
              (json['confidence'] as Map<String, dynamic>)['question']
                  as Map<String, dynamic>,
            ),
      confidenceAnswer = json['confidence'] == null
          ? null
          : (json['confidence'] as Map<String, dynamic>)['selectedOptionId']
                as String,
      review = json['review'] == null
          ? null
          : DiagnosticReview.fromJson(json['review'] as Map<String, dynamic>) {
    if (!['active', 'completed'].contains(json['status']) ||
        total <= 0 ||
        answered < 0 ||
        answered > total ||
        (!completed && question == null && confidenceQuestion == null)) {
      throw const FormatException('Invalid diagnostic progress');
    }
  }
  final bool completed;
  final int answered;
  final int total;
  final DiagnosticQuestion? question;
  final DiagnosticQuestion? confidenceQuestion;
  final String? confidenceAnswer;
  final DiagnosticReview? review;
}

class DiagnosticFailure implements Exception {
  const DiagnosticFailure(this.message);
  final String message;
}

File: apps/api/src/modules/diagnostic/domain/diagnostic.ts
export const CONFIDENCE_VALUES = [
  'guessing',
  'somewhat_sure',
  'very_sure',
] as const;
export type Confidence = (typeof CONFIDENCE_VALUES)[number];
export type QuestionCategory =
  | 'scenario_judgment'
  | 'predict_outcome'
  | 'spot_the_bug'
  | 'better_approach'
  | 'conceptual_reasoning';
export type Concept = Readonly<{ id: string; label: string }>;
export interface DiagnosticQuestion {
  readonly id: string;
  readonly category: QuestionCategory;
  readonly interactionType: 'single_choice';
  readonly skill: Concept;
  readonly concept: Concept;
  readonly difficulty: string;
  readonly prompt: string;
  readonly context: string | null;
  readonly code: string | null;
  readonly options: readonly Readonly<{ id: string; label: string }>[];
  readonly confidenceRequested: boolean;
  readonly correctOptionId: string;
  readonly explanation: string;
  readonly keyIdea: string;
}
export interface DiagnosticDefinition {
  readonly id: string;
  readonly version: string;
  readonly questions: readonly DiagnosticQuestion[];
}
export interface DiagnosticSession {
  id: string;
  diagnosticId: string;
  definitionVersion: string;
  status: 'active' | 'completed';
  startedAt: Date;
  completedAt: Date | null;
}
export interface DiagnosticAttempt {
  questionId: string;
  skillId: string;
  conceptId: string;
  difficultyId: string;
  interactionType: string;
  selectedOptionId: string;
  correct: boolean;
  responseDurationMs: number | null;
  confidence: Confidence | null;
  answeredAt: Date;
}
export type DiagnosticFailureCode =
  | 'PROFILE_INCOMPLETE'
  | 'DIAGNOSTIC_UNAVAILABLE'
  | 'DIAGNOSTIC_NOT_STARTED'
  | 'DIAGNOSTIC_COMPLETED'
  | 'INVALID_DIAGNOSTIC_ANSWER'
  | 'DIAGNOSTIC_ANSWER_CONFLICT'
  | 'DIAGNOSTIC_QUESTION_OUT_OF_ORDER'
  | 'INVALID_DIAGNOSTIC_CONFIDENCE';
export class DiagnosticFailure extends Error {
  constructor(readonly code: DiagnosticFailureCode) {
    super(code);
  }
}
export function nextQuestion(
  definition: DiagnosticDefinition,
  attempts: DiagnosticAttempt[],
): DiagnosticQuestion | undefined {
  return definition.questions.find(
    (question) =>
      !attempts.some((attempt) => attempt.questionId === question.id),
  );
}
export function pendingConfidence(
  definition: DiagnosticDefinition,
  attempts: DiagnosticAttempt[],
): DiagnosticAttempt | undefined {
  return attempts.find(
    (attempt) =>
      attempt.confidence === null &&
      definition.questions.some(
        (question) =>
          question.id === attempt.questionId && question.confidenceRequested,
      ),
  );
}
export function evaluateAnswer(
  question: DiagnosticQuestion,
  selectedOptionId: string,
  duration?: number,
): Omit<DiagnosticAttempt, 'answeredAt'> {
  if (
    !question.options.some((option) => option.id === selectedOptionId) ||
    (duration !== undefined &&
      (!Number.isInteger(duration) || duration < 0 || duration > 86400000))
  ) {
    throw new DiagnosticFailure('INVALID_DIAGNOSTIC_ANSWER');
  }
  return {
    questionId: question.id,
    skillId: question.skill.id,
    conceptId: question.concept.id,
    difficultyId: question.difficulty,
    interactionType: question.interactionType,
    selectedOptionId,
    correct: selectedOptionId === question.correctOptionId,
    responseDurationMs: duration ?? null,
    confidence: null,
  };
}
